### PR TITLE
Noita: implement new game

### PR DESCRIPTION
**Pull request by heinermann**
## What is this fixing or adding?
Adding Noita (PC game) to Archipelago.

## How was this tested?
Manual verification and existing unit tests. No automated testing was added. 2 beta tests.

## Additional Notes
- Old PR [here](https://github.com/ArchipelagoMW/Archipelago/pull/989).
- Will the character ä cause any problems for other games?
- Went for a slim `__init__.py`.
- Please point out if anything is unclear at first glance.
- Open to test suggestions since I haven't familiarized myself with the test framework yet.

### CODE REVIEW

File: worlds/noita/Items.py
Comment (heinermann) [inline worlds/noita/Items.py:4]: Might be clearer to use fully qualified `Options.whatever` below, same elsewhere.

File: worlds/noita/Locations.py
Comment (heinermann) [inline worlds/noita/Locations.py:47]: We can make the following changes to improve readability for outsiders:

`"hc"` -> `"chest"`
`"peds"` -> `"pedestal"`

File: worlds/noita/Rules.py
Comment (heinermann) [inline worlds/noita/Rules.py:28]: Gold (10) and Gold (50) are no longer in the item list and should be removed.

File: worlds/noita/Rules.py
Comment (heinermann) [inline worlds/noita/Rules.py:65]: should just use the location directly, I don't see why the loop is needed?

File: worlds/noita/__init__.py
Comment (heinermann) [inline worlds/noita/__init__.py:4]: Import not needed?

**Comment by ThePhar** [comment]
Talked with the devs in the Noita Implementation forum thread. Going to wait on doing a full review until some Rules rework to add some artificial gating for better spoiler output and progression balancing.

File: worlds/noita/Items.py
Comment (ThePhar) [inline worlds/noita/Items.py:44]:
```suggestion
        itempool.append("Perk (Spatial Awareness)")
```

File: worlds/noita/Items.py
Comment (ThePhar) [inline worlds/noita/Items.py:37]: No need to check this with a separate `if` if the previous statement was true. :P

```suggestion
    elif vic == Options.VictoryCondition.option_peaceful_ending:
        orb_count = 33
```

File: worlds/noita/Rules.py
Comment (ThePhar) [inline worlds/noita/Rules.py:132]: You'll also want to set an access rule on the "Victory" location for the conditions that require 11 or 33 orbs. Something like so, although you can move it wherever you want.

```suggestion
        forbid_items_at_location(world, "Toveri", {"Perk (Spatial Awareness)"}, player)

    # Set rule for victory condition.
    vic = world.victory_condition[player].value
    victory_location = world.get_location("Victory", player)
    if vic == Options.VictoryCondition.option_pure_ending:
        victory_location.access_rule = lambda state: has_orb_count(state, player, 11)
    elif vic == Options.VictoryCondition.option_peaceful_ending:
        victory_location.access_rule = lambda state: has_orb_count(state, player, 33)

```

Comment (ThePhar) [reply worlds/noita/Rules.py:132]: Otherwise, calculating playthrough/win condition will not list all the required orbs for victory, only to get into each section. As it thinks you only need enough for the most expensive location.

**Review by ThePhar — CHANGES_REQUESTED** [review]
Overall looks like it works pretty well, all tests pass just a few minor performance things to tweak and adding an access rule to the "Victory" location so the correct number of orbs is calculated when determining play through and beat-ability.

After these tweaks are made, would like to see a couple beta tests just to make sure, but looks almost ready! :)

File: worlds/noita/Items.py
Comment (ThePhar) [reply worlds/noita/Items.py:41]: This also removes the `for i in range(orb_count):` line too, since it's done in one line with a comprehension.

**Comment by heinermann** [comment]
We should describe what the Sampo is to avoid confusion.

File: worlds/noita/__init__.py
Comment (heinermann) [reply worlds/noita/__init__.py:43]: It's in the `RoomInfo` packet

Comment (ThePhar) [reply worlds/noita/__init__.py:43]: You are correct, I misremembered.

File: worlds/noita/Locations.py
Comment (heinermann) [inline worlds/noita/Locations.py:29]: We might want to revisit this to close the large gaps , we have ~669 locations, not 2300 locations

File: worlds/noita/Items.py
Comment (heinermann) [reply worlds/noita/Items.py:6]: Decided against it

File: worlds/noita/Locations.py
Comment (ScipioWright) [reply worlds/noita/Locations.py:29]: Yeah... I've been wanting to put off doing this, but probably will after the sync today or tomorrow at some point.

**Comment by ThePhar** [comment]
Just posting here so other core maintainers know whats up:

Since you all were telling me you wanted to make some tweaks, just give me the go-ahead and assuming everything is good once ready, I'll review this again. :) Beta test looked promising.

**Comment by ScipioWright** [comment]
Will let you know! We wanted to do some more game-side updates before the full release, since while the ap-side stuff itself is done (except for squishing the item and location IDs so we don't have such a big gap), we want the issues found during the tests fixed before release, so we aren't releasing something buggy (even though that seems to be the norm in gaming as a whole nowadays...)

**Comment by ScipioWright** [comment]
This is ready for final review now. We have made the changes we wanted to make, and believe this is in a good place.

File: worlds/noita/__init__.py
Comment (ThePhar) [inline worlds/noita/__init__.py:29]: Considering you want to get this merged, let's also add `data_version = 1` so the Data Package isn't in custom mode anymore.

```suggestion
    data_version = 1
```

File: worlds/noita/__init__.py
Comment (ThePhar) [inline worlds/noita/__init__.py:38]: Just a small code optimization, since you only use this function in `def fill_slot_data`, you could rewrite remove this method and write `fill_slot_data` like so:

```py
    # Returned items will be sent over to the client
    def fill_slot_data(self):
        return {name: getattr(self.multiworld, name)[self.player].value for name in self.option_definitions}
```

File: worlds/noita/__init__.py
Comment (ThePhar) [inline worlds/noita/__init__.py:28]: Probably not necessary to set since routing for each location does not change for each generation and saves some output in spoiler output mode 3.

**Review by ThePhar — CHANGES_REQUESTED** [review]
LGTM just a couple tweaks to be made.

**Review by ThePhar — APPROVED** [review]

**Pull request merged**
